### ISSUE TITLE
Outgoing Invalid in yellow folders; New messages Valid, replies Invalid

### ISSUE THREAD

**Issue by Tom-Usher** [lieser/dkim_verifier#111]
Thunderbird 52.7.0 (32-bit)
DKIM Verifier 1.7.0

Everything appears to be working pretty well with the exception that my outgoing messages are showing as invalid but only when they are in a yellow folder. This appears to happen for me only when Options > General > Policy > "Automatically add rules ..." is selected.

The above would appear to be an extension issue.

There's another hitch, which may or may not be an email-server issue.

Results for new messages and replies are different regardless of the folder. New messages show valid at the recipient's end but invalid for replies. The recipient in this case is also using Thunderbird 52.7.0 (32-bit) and DKIM Verifier 1.7.0.

All of this is happening via email running through the same domain/same email server.

This block is for an offending reply message :
Authentication-Results: vade-backend14.dreamhost.com; dkim=fail
	reason="verification failed; unprotected key"
	header.d=propertypak.com header.i=@propertypak.com
	header.b=dq95Psvm; dkim-adsp=none (unprotected policy);
	dkim-atps=neutral

This block is for a new message:
Authentication-Results: vade-backend14.dreamhost.com; dkim=pass
	reason="1024-bit key; unprotected key"
	header.d=propertypak.com header.i=@propertypak.com
	header.b=j71AcYVw; dkim-adsp=pass; dkim-atps=neutral

I do hope you can duplicate these issues. The extension is quite useful and important. We greatly appreciate your time and effort and hope this submission helps you keep this extension free.

If you can't duplicate them and if you believe either issue somehow falls with the email-server config, I would appreciate any information you'd suggest I pass on to the host.

**Comment by lieser** [comment]
### 1. my outgoing messages are showing as invalid:

That is the exact error message you are getting? Is it `Should be signed by ...` or a different error? If it is a different error, this is expected behavior. Only the `Should be signed by ...` error is treated special for outgoing messages.

In case it is `Should be signed by ...`:
For some reason the detection that this is an outgoing e-mail seems to fail. Please make sure that there exist an identity in Thunderbird for the Account that is identically to the sending e-mail address.

About the `Automatically add rules ...`:
The `Should be signed by ...` error will only be shown if there exist an active rule that tells the add-on that the e-mail should be signed. The `Automatically add rules ...` will automatically add such rules. You can manually add a rule for the e-mail address to overwrite the automatically added one.

If you haven't already, you should have a look at https://github.com/lieser/dkim_verifier/wiki/Options#policy-tab and https://github.com/lieser/dkim_verifier/wiki/Sign-rules (at the end there is a small example for *Disable "Should be signed by ..." error*. You can also add the exception via the other menu in the e-mail header)

### 2. Results for new messages and replies are different regardless of the folder

You posted some `Authentication-Results` headers (ARH), there on is saying `dkim=fail`. Note that this are not written by the add-on. They can only be read by the add-on.
If you currently set the add-on to read the ARH header, you should temporarily disable this so that the add-on does a local verification. If the add-on is coming to the same result as the e-mail server in the ARH, it is very likely an issue with the signature creation.

**Comment by Tom-Usher** [comment]
Thank you for your reply. I understand everything you've said.

I'll reply separately concerning each issue. This reply will concern only: "1. my outgoing messages are showing as invalid:".

The exact wording by your extension is: DKIM Invalid (Should be signed by propertypak.com).

I verified again that enabling Options > General > Policy > "Automatically add rules ..." and leaving the automatically created rule for the domain in the "Signers rules" results in the above error. Deleting the rule for the domain results in "DKIM No Signature."

Therefore, you appear to be correct that "... the detection that this is an outgoing e-mail seems to fail."

You wrote: "Please make sure that there exist an identity in Thunderbird for the Account that is identically to the sending e-mail address." I've been testing back and forth via accounts that belong to me and all of which are setup in my Thunderbird.

Were you not able to duplicate this issue? My associate with the same setup experiences the exact same problem.

Thank you.

**Comment by Tom-Usher** [comment]
Please keep in mind that this "outgoing" issue only occurs when the outgoing messages are in a yellow folder (a folder created by me, not a default Thunderbird "Sent" folder or any Thunderbird default folder). Therefore the detection needs to cover more than what it currently covers. It needs to handle non-default folders, which I'm sure you understand. I have no intention of insulting your intelligence, which is obviously high.

**Comment by Tom-Usher** [comment]
Concerning: "2. Results for new messages and replies are different regardless of the folder":

I have been dealing with the email host on this aspect and am unsure at this point what if anything they've done on their end. The problem appeared to clear up yesterday. Your extension appears to be functioning properly concerning both new and reply messages.

Unless the issue returns, I'll consider it a one-time hiccup on the host's part. If it does return, I'll work on it with the host, as I'm confident now that it was not, and won't be, an issue with your extension.

Thank you for your patience and assistance/explanation.

**Comment by Tom-Usher** [comment]
Update: Concerning: "2. Results for new messages and replies are different regardless of the folder":

I'm adding this just in case anyone else is experiencing the same issue.

The problem returned yesterday.

I've sent email headers to the email host so they may escalate it.

If they report anything back to me that has any bearing on this extension, I'll post another update with that info.

Thanks again.

**Comment by lieser** [comment]
### 1. my outgoing messages are showing as invalid:

> Were you not able to duplicate this issue? My associate with the same setup experiences the exact same problem.

I'm not able to duplicate the issue. I just tested again with the exact same Thunderbird version 52.7.0
Would you be willing to install a custom version of the extension with more debug information?

> Please keep in mind that this "outgoing" issue only occurs when the outgoing messages are in a yellow folder (a folder created by me, not a default Thunderbird "Sent" folder or any Thunderbird default folder).

The `Sent` folder is treated differently. All messages in the folder are assumed outgoing by the add-on (and are therefore never required to be signed). This does not apply to other default folders like `Inbox` or `Trash`. They should show the same behavior as over normal (yellow) folders.  
Are the folders you are talking about maybe folders under the `Locals Folder` account? As this account has no concept of Identities, the current outgoing message detection will not work.

**Comment by Tom-Usher** [comment]
Philippe,

You asked the exactly correct question.

I just finished testing:

I created a new folder under the sending account (not under Local Folders). I put a sent message in it, and the extension reported correctly "No Signature."

I simply then dragged that folder to a different account (not under Local Folders), and Thunderbird duplicated it rather than moving it and also duplicated the single sent message right within it. For that duplicate, the extension reported "Invalid (Should be signed by propertypak.com)."

No matter which account's default Sent box an outgoing message is in, the extension reports "No Signature."

If the same outgoing message is placed in any folder (default or manually created) under the outgoing account, the extension correctly reports "No Signature."

So, any user of the extension who has created lots of folders under Local Folders and has created lots of filters for sent messages to be automatically moved to any of those created folders, as I have, will either have to have Options > General > Policy > "Automatically add rules ..." disabled and no rule for the applicable domain or just ignore "Invalid (Should be signed by xxx.xxx)."

I don't know what bearing, if any, it may have on this issue, but I have "Only if the From address is in the SDID" enabled and set to "for base domain."

Are you positive you can't duplicate the problem?

If you are not able to duplicate it, even though two of us independently have the same problem on opposite sides of the country here, I'm not sure what to think.

I have helped developers by installing alpha versions to test patches and am willing to do that in this case, but I would like you to first please try one more time on your end to duplicate the error.

As I know trains-of-thought can be hard to follow and that I appreciate redundant clarity (when not taken to extreme) in complicated matters, let me say that you need to be sure there on your end that you have the Options > General > Policy > "Automatically add rules ..." enabled and that a rule exists for the domain you use to test outgoing messages to see if you can duplicate this problem two of us here are experiencing.

Thank you, Philippe, for your efforts on this.

**Comment by lieser** [comment]
> Are you positive you can't duplicate the problem?

There seems to be have a misunderstanding. Like I wrote in my last comment, the current automatic detection for outgoing messages will not work for the `Locals Folder` account, resulting in the behavior you observe.
As until your last comment I assumed you are talking about folders under POP3/IMAP accounts, I meant I can not reproduce it for this account types (POP3/IMAP).

> [...], will either have to have Options > General > Policy > "Automatically add rules ..." disabled and no rule for the applicable domain or just ignore "Invalid (Should be signed by xxx.xxx)."

This is not true. But you have to manually add sign rules for your own e-mail addresses that state that explicitly e-mails from them must not be signed, regardless of that over rules (with lower priority) may say.
Have you already tried adding such rules?

As a reminder, from my first comment:
> If you haven't already, you should have a look at https://github.com/lieser/dkim_verifier/wiki/Options#policy-tab and https://github.com/lieser/dkim_verifier/wiki/Sign-rules (at the end there is a small example for Disable "Should be signed by ..." error. You can also add the exception via the other menu in the e-mail header)

**Comment by Tom-Usher** [comment]
There was a misunderstanding because it did not occur to me that people manually create folders under accounts rather than under Local Folders. However, I did fully understand your message to which I was replying. The distinction wasn't lost on me. I could have been clearer. I was tempted yesterday to return to this thread to acknowledge that I was taking from your statements that everything I've described is known and expected behavior.

As for manually adding a signed rule, I was time-constrained (and still am) but will read the linked materials and work with creating a rule that will work properly in all desired ways. From your last message, I feel confident that reaching that state will be fairly simple. Until you asked that exact question concerning the Local Folders, I just wasn't sure whether we weren't talking past each other (https://en.wikipedia.org/wiki/Talking_past_each_other) and held off trying to do other things concerning the issue until you expressly raised the "Local Folders" aspect.

So, my apologies for not simply diving into the linked material and adding a rule manually.

Thanks for your patience and thoroughness.

If I have no problem creating the rule and if the rule works as advertised, I'll not add more to this thread.

Lastly, the other issue in my very first comment did turn out to be an issue at the host's server(s). They're working on it now.

**Comment by Tom-Usher** [comment]
Let me add that I had read both linked-pages when making setting decisions concerning the extension and before my first comment but will re-read them now, especially now that I have my head wrapped around DKIM. Your extension, which I only installed a few days before the issue showed up to me is my first experience ever dealing with DMARC.

**Comment by lieser** [comment]
As you haven't responded yet, I assume manually adding the rule worked for you. Note that I created a separated issue (#114) to extend the outgoing detection to also work in the Local Folder.
